participant/app/utils/s3.server.ts

```diff
@@ -0,0 +1,198 @@
+    }
+    throw new Error(`Unable to complete head request for ${key}: ${error}`);
+  }
+};
+
+export const getURLFromS3 = async (
+  key: string,
+  duration?: number
+): Promise<string | undefined> => {
+  await ensureBucketExists(s3Connection);
+  const expiresIn = duration || 3600;
+  const command = new GetObjectCommand({
+    Key: key,
+    Bucket: BUCKET,
+  });
+  try {
+    const signedUrl = await getSignedUrl(s3Connection, command, {
+      expiresIn: expiresIn,
+    });
+    return signedUrl;
+  } catch (error) {
+    if (error instanceof NotFound) {
+      return undefined;
+    }
+    throw new Error(`Unable to get URL for ${key}: ${error}`);
+  }
+};
+
+export const deleteFileFromS3 = async (key: string) => {
+  await ensureBucketExists(s3Connection);
+  const command = new DeleteObjectCommand({
+    Key: key,
+    Bucket: BUCKET,
+  });
+  try {
+    await s3Connection.send(command);
+    console.log(`🗑️  Deleted ${key}`);
+  } catch (error) {
+    // If it doesn't exist, it seems like an OK outcome for a delete request
+    if (error! instanceof NotFound) {
+      return;
+    } else {
+      throw new Error(`Unable to delete ${key}: ${error}`);
+    }
+  }
+};
+
+export const checkFile = async (key: string): Promise<FileCheckResult> => {
+  const fileSize = await headFilesizeFromS3(key);
+  if (!fileSize) {
+    return { error: "notFound" };
+  } else if (!validSize(fileSize)) {
+    return { error: "invalidSize", size: fileSize };
+  }
+
+  const startOfFile = await readFileHeadFromS3(key);
+  if (!startOfFile) {
+    console.error(`❌ Unable to get head of file for ${key}`);
+    return { error: "cannotRead", size: fileSize };
+  }
+  const fileType = await fileTypeFromBuffer(startOfFile);
+  if (!fileType) {
+    console.error(`❌ Unable to determine filetype for ${key}`);
+    return { error: "cannotType", size: fileSize };
+  }
+  if (fileType.mime.includes("image") || fileType.mime == "application/pdf") {
+    return { mimeType: fileType.mime, size: fileSize };
+  }
+  return { mimeType: fileType.mime, error: "invalidType", size: fileSize };
+};
+
+// Thank you 🙏🏻 to https://github.com/remix-run/examples/issues/163
+const uploadStream = ({ Key }: Pick<PutObjectCommandInput, "Key">) => {
+  const pass = new PassThrough();
+  return {
+    writeStream: pass,
+    promise: new Upload({
+      params: {
+        Body: pass,
+        Bucket: BUCKET,
+        Key,
+      },
+      client: s3Connection,
+    }).done(),
+  };
+};
+
+export async function uploadStreamToS3(data: any, filename: string) {
+  await ensureBucketExists(s3Connection);
+  const stream = uploadStream({
+    Key: filename,
+  });
+  await writeAsyncIterableToWritable(data, stream.writeStream);
+  const file = await stream.promise;
+  if ("Location" in file) {
+    if (PATHSTYLE) {
+      // Workaround to ensure we're logging valid URLs in dev
+      // The port number is missing from file.Location
+      return `${ENDPOINT_URL}/${file.Bucket}/${file.Key}`;
```

Comment:
Does this mean that `ENDPOINT_URL` is only used for local development? Not on AWS?

Reply:
100%.
The clients all have default connection patterns for connectin in S3; endpoint url is the unusual behavior